pkg/cli/server/server.go

```diff
@@ -374,15 +375,16 @@ func run(app *cli.Context, cfg *cmds.Server, leaderControllers server.CustomCont
 			}
 			serverConfig.ControlConfig.ClusterDNSs = append(serverConfig.ControlConfig.ClusterDNSs, parsed)
 		}
-		// Set ClusterDNS to the first IPv4 address, for legacy clients
-		// unless only IPv6 range given
-		clusterDNS, _, _, err := util.GetFirstIP(serverConfig.ControlConfig.ClusterDNSs)
-		if err != nil {
-			return errors.Wrap(err, "cannot configure IPv4/IPv6 cluster-dns address")
-		}
-		serverConfig.ControlConfig.ClusterDNS = clusterDNS
 	}
 
+	// Set ClusterDNS to the first IPv4 address, for legacy clients
+	// unless only IPv6 range given
+	clusterDNS, _, _, err := util.GetFirstIP(serverConfig.ControlConfig.ClusterDNSs)
```

Comment:
Is this correct? I think Kubernetes might insist that the primary clusterIP for the service respect the ip family order for the cluster.

Reply:
As of today, all our code is handling this situation prioritizing ipv4 regardless of the order. I am preparing another PR to change this. In any case, it is not clear that Kubernetes mandates anything in this respect, in fact, when they describe dual-stack, IPv4 is always the first in all their examples. We can discuss further in my other PR

Reply:
BTW, I am creating a different PR because verifying the order to decide what IPFamily gets prioritized requires changes in several parts of the code

Reply:
Will this break if I set `--service-cidr 2001:cafe:42:1::/112,10.43.0.0/16 --cluster-dns 2001:cafe:42:1::a,10.43.0.10`?

If the kube-dns service in the `coredns.yaml` will be rendered as:

```yaml
clusterIP: 10.43.0.10
clusterIPs: [2001:cafe:42:1::a, 10.43.0.10]
ipFamilyPolicy: PreferDualStack
```

Then I think things will go wrong, as the [`ServiceSpec`](https://kubernetes.io/docs/reference/kubernetes-api/service-resources/service-v1/#ServiceSpec) says:

> - clusterIPs ([]string)
> ...
> If this field is not specified, it will be initialized from the clusterIP field. If this field is specified, clients must ensure that clusterIPs[0] and clusterIP have the same value.
> ...

`clusterIP` and `clusterIPs[0]` must always be identical.

Reply:
As of today, all our code is expecting to define dual-stack as stated in the Kubernetes documentation:
https://kubernetes.io/docs/concepts/services-networking/dual-stack/#configure-ipv4-ipv6-dual-stack
, i.e. it expects the order to be `<IPv4 CIDR>,<IPv6 CIDR>`. K3s is currently not ready for configs like `--service-cidr 2001:cafe:42:1::/112,10.43.0.0/16`. You are right, your config will not work, but there are other things that will not work either without some workarounds: e.g. kube-api or netpol controller. I am working on another PR to provide support for `<IPv6 CIDR>,<IPv4 CIDR>` type of configs. I'll add you to that PR as reviewer and we can address this problem over there

Reply:
It actually has nothing to do with the order of `--service-cidr`. `--service-cidr 10.43.0.0/16,2001:cafe:42:1::/112 --cluster-dns 2001:cafe:42:1::a,10.43.0.10` will lead to the same result.

Kind of feels like there should be a check here to make sure that `cmds.ServerConfig.ClusterDNS` always matches `cmds.ServerConfig.ServiceCIDR` (within the ranges, having the same families and family order no matter which family takes precedence). I think k3s should throw a fatal to prevent the above case from happening.

Reply:
Correction: `cmds.ServerConfig.ClusterDNS` should always matches `cmds.ServerConfig.ServiceCIDR` __when__ `len(cmds.ServerConfig.ClusterDNS) == len(cmds.ServerConfig.ServiceCIDR)`.

The case of `--service-cidr 10.43.0.0/16,2001:cafe:42:1::/112 --cluster-dns 10.43.0.10` should still be allowed.

Reply:
I agree with all of the above. If there are places where we are forcing or expecting IPv4 as the primary AF on dual-stack clusters, we should work towards eliminating those.

On the other hand, if someone changes the cluster-dns setting after the cluster has already been started, I'm not sure how hard we need to go our of our way to handle changes not allowed by Kubernetes itself.

Reply:
In my opinion:
```
--service-cidr 10.43.0.0/16,2001:cafe:42:1::/112 --cluster-dns 2001:cafe:42:1::a,10.43.0.10
```
is currently not acceptable because cluster-dns is using the non-supported order: `<IPv6><IPv4>`. 

Let me finish up the other PR and we can discuss `<IPv6><IPv4>` over there. I think I can have it in a few hours

Reply:
https://github.com/k3s-io/k3s/pull/8460 <== When using this PR, the problem will not exist anymore (I think)